### ISSUE TITLE
Duration Discriminator problem

### ISSUE THREAD

**Issue by yiwei0730** [p0p4k/vits2_pytorch#68]
In train_ms.py line 408-411，The duration Discriminator Net  in the last two input , Is the order reversed? 
I saw the other vits2 code order reversed, but  logw_ is the MAS result logw is the dp result, I think this is correct. 
Can you help me answer it?
# Duration Discriminator
            if net_dur_disc is not None:
                y_dur_hat_r, y_dur_hat_g = net_dur_disc(
                    hidden_x.detach(), x_mask.detach(), logw_.detach(), logw.detach()
                )
->      
# another Duration Discriminator
           if net_dur_disc is not None:
                y_dur_hat_r, y_dur_hat_g = net_dur_disc(
                    hidden_x.detach(), x_mask.detach(), logw.detach(), logw_.detach()
                )

**Comment by p0p4k** [comment]
`logw` is fake (coming from dp); `logw_` is 'real' (coming from MAS during training only)
Discriminator takes in real, fake gives real's_prob (prob that the real is real), fake's_prob (prob that fake is real).

**Comment by yiwei0730** [comment]
Yes, you are right. I see the another Duration Discriminator in bert-vits2 repo, so its make me confuse. 
Thanks for your reply!!

**Comment by p0p4k** [comment]
They all are my copies 🤣 it is when I had made mistakes at initial commit.

**Comment by deyituo** [comment]
hh, I use bert-vits2 to train base model  on genshin+starrail dataset and find that loss_dur_gen is 40+ and to cause the loss nan. But when training on paimon, the loss is smaller. 
Finally I found here, maybe this is the issue.

**Comment by yiwei0730** [comment]
> hh, I use bert-vits2 to train base model on genshin+starrail dataset and find that loss_dur_gen is 40+ and to cause the loss nan. But when training on paimon, the loss is smaller. Finally I found here, maybe this is the issue.

yes they make the prediction to be the real one. There system is perfect, but still have small bugs.

**Comment by deyituo** [comment]
I find that the dur discriminator loss is still large, em...

**Comment by deyituo** [comment]
vits2's paper says that just train for 30k steps, en...

**Comment by JohnHerry** [comment]
> > hh, I use bert-vits2 to train base model on genshin+starrail dataset and find that loss_dur_gen is 40+ and to cause the loss nan. But when training on paimon, the loss is smaller. Finally I found here, maybe this is the issue.
> 
> yes they make the prediction to be the real one. There system is perfect, but still have small bugs.

Hi，yiwei0730， can you share me other bugs in the Bert-VITS2 please?  I had experienced wrong phoneme pronunciation when using it.  I had considered that the BERT feature had been disturbing the phoneme features so that some phonemes are wrongly uttered.  But now you point out that there are some bugs. so I am interested in this.

**Comment by yiwei0730** [comment]
they use SDP/DP ratio to train, but i'm not sure this is correct. But I can choose the duration ratio in inference, maybe is a good idea. and also in 2.1 they still didn't found the logw_ problem.
There preprocess if you are focus on chinese, maybe have some wrong in preprocess.
I'm not sure if any problem in the code inside the model, i'm not read all the code yet.

**Comment by JohnHerry** [comment]
> they use SDP/DP ratio to train, but i'm not sure this is correct. But I can choose the duration ratio in inference, maybe is a good idea. and also in 2.1 they still didn't found the logw_ problem. There preprocess if you are focus on chinese, maybe have some wrong in preprocess. I'm not sure if any problem in the code inside the model, i'm not read all the code yet.

I had tried the v1.0.1 because I do not need multi-lingual support.  Yes I have the some doubt that they are training the SDP and DP together by sum them up with ratio. I do not know if that is the reason of those bad pronunciations. It seems that they are trying to add new features to the project to do more, but no model optimization and improvement at all.
I am now trying to fix the logw_ problem and disable the  add_blank config in Bert-VITS2, hope that will be helpfull.

**Comment by deyituo** [comment]
@JohnHerry Maybe you should start with at least v2.0.2-fix, v1.x is not good enough in quality

**Comment by deyituo** [comment]
After I remove the duration discriminator to train multi/single corpus, the loss seems to be more like other projects' logs.

gensin+starrail
![image](https://github.com/p0p4k/vits2_pytorch/assets/19730584/78f9321f-fc3a-4adf-aeee-0dbf8d1e99cf)
![image](https://github.com/p0p4k/vits2_pytorch/assets/19730584/365cfe35-a14f-426d-9c8a-283cd7917c9b)

paimon:
![image](https://github.com/p0p4k/vits2_pytorch/assets/19730584/6a922fdc-4972-49ec-8998-1b2bf582b2a0)
![image](https://github.com/p0p4k/vits2_pytorch/assets/19730584/79fdd99a-49a1-4147-872b-426da3952bc4)

**Comment by JohnHerry** [comment]
> @JohnHerry Maybe you should start with at least v2.0.2-fix, v1.x is not good enough in quality

When I start, It newest  version is v2.0.1,  I had compared most of the  import files:  commons.py  losses.py  models.py modules.py, transforms.py,  and there is no important difference with v1.0.1,  the changes are about multi-lingual support which I am not care about. so I used the v1.0.1 as my base.

**Comment by JohnHerry** [comment]
> After I remove the duration discriminator to train multi/single corpus, the loss seems to be more like other projects' logs.
> 
> gensin+starrail ![image](https://private-user-images.githubusercontent.com/19730584/287894910-78f9321f-fc3a-4adf-aeee-0dbf8d1e99cf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NDQ2NTIsIm5iZiI6MTcwMTc0NDM1MiwicGF0aCI6Ii8xOTczMDU4NC8yODc4OTQ5MTAtNzhmOTMyMWYtZmMzYS00YWRmLWFlZWUtMGRiZjhkMWU5OWNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDAyNDU1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxYmRlYmU2Y2VmNDdiYzY4ZDljMDZmNWQyMmU2YmVjZGJmOGExYzdmNmJiYzM4OTQ4MjZlODdmNmRkZGQzNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jX7cgsEmq015GJbSJuxRUbHP0kjVeO3eqvu20GFRThI) ![image](https://private-user-images.githubusercontent.com/19730584/287894936-365cfe35-a14f-426d-9c8a-283cd7917c9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NDQ2NTIsIm5iZiI6MTcwMTc0NDM1MiwicGF0aCI6Ii8xOTczMDU4NC8yODc4OTQ5MzYtMzY1Y2ZlMzUtYTE0Zi00MjZkLTljOGEtMjgzY2Q3OTE3YzliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDAyNDU1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE5N2RiYTA3ZDMwMDVkM2NiZjJkNGE2ODgxNmQ2NTdhZjFiOGNlMzU3MmNiMWFhOTI1ZjNhYjU5OGNiZDVlZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Uz8s66eTIbvjFBlpt62HzifpKKPfpBZKi8cz7G5o0Do)
> 
> paimon: ![image](https://private-user-images.githubusercontent.com/19730584/287895121-6a922fdc-4972-49ec-8998-1b2bf582b2a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NDQ2NTIsIm5iZiI6MTcwMTc0NDM1MiwicGF0aCI6Ii8xOTczMDU4NC8yODc4OTUxMjEtNmE5MjJmZGMtNDk3Mi00OWVjLTg5OTgtMWIyYmY1ODJiMmEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDAyNDU1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzYWYxNmE3NmMxNDBkYmRhYTExNjYxZmExNDY2MGRlZTM5YjhiYjRjOGQ0OTE1MThkOTE2ZGJmMWUwZjQ0ZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tjC_baZLMgpOHc-eIP7tCVyUOwqOyTFj9w1sBWZzi3c) ![image](https://private-user-images.githubusercontent.com/19730584/287895152-79fdd99a-49a1-4147-872b-426da3952bc4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NDQ2NTIsIm5iZiI6MTcwMTc0NDM1MiwicGF0aCI6Ii8xOTczMDU4NC8yODc4OTUxNTItNzlmZGQ5OWEtNDlhMS00MTQ3LTg3MmItNDI2ZGEzOTUyYmM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDAyNDU1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkYWVhMThiOTUzZWYyMzA4YzVjNTU5OTJlZmU1ZDEyNWIxZGViMWY1NjExODM4NDViZTMxYmZjYzUwNTE4MjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1wOsaVbgtKh78XrZIGNj2MHVFcRLWedeSt5g4bFJ9qc)

Yes, remove the DD will make the training process more stable, but as an important point in the vits2 paper, I did not disable this,  I follow the paper to train the acount model for about 700-800K steps and then enable DD and train them together,   after the DD traning, the bad pronunciation problem get better, but still exist.

**Comment by p0p4k** [comment]
@JohnHerry are you using sdp and dur_disc OR deterministic dp and dur_disc?

**Comment by JohnHerry** [comment]
> @JohnHerry are you using sdp and dur_disc OR deterministic dp and dur_disc?

I was using Bert-VITS2 v1.0.1, it makes DP by   DP = (  sdp * sdp_ratio +  dp * (1-sdp_ratio)), Yes, it training and inference with those two DPs together,  strange but usfull.    sdp_ratio =1 make the result more natural,   sdp_raio=0 make the result more stable.  as we know, the SDP will sometimes generate very bad result.

**Comment by p0p4k** [comment]
I see, interesting approach. Thanks for sharing, I think `sdp_ratio` can be made into a learnable parameter during training itself.

**Comment by JohnHerry** [comment]
> I see, interesting approach. Thanks for sharing, I think `sdp_ratio` can be made into a learnable parameter during training itself.

Oh,  wonderful if it is  made learnable, but I think it is not an easy task. In my understanding, some good jobs like this vits2 and StyleTTS2, are using exactly tow-stage training.  The first stage they train a roughly usable acounstic model, and the seoncd stage is to train a better model. VITS2 firstly learn 800K steps, and then learn a better DP with the DD; the StyleTTS2 first train many epoches on the acounstic model and then the DDPM based Speech Style predictor.  As to the sdp_ratio, I think, if it is trainable, then this traning should start at somepoint that the SDP and DP are roughly usable, then the training pipeline may become more complex.
